### ISSUE TITLE
2-part navigation required before Views in Child Regions can be cleaned up [Bug] 

### ISSUE THREAD

**Issue by james1301** [PrismLibrary/Prism#2608]
### Description

A view with child views is not cleaned up when the parent view is closed. I am using an altered ClearChildViewsRegionBehaviour that always (Doesn't need to set the attached property) clears child views and does this by Deactivating the view (instead of just setting the region manager to null), this replaces the default behavior.

### Steps to Reproduce

1. Run application
2. Click to Navigate to View with Child Views button
3. Click the close button
4. You will see a message box that says "View with Children Disposed"
5. Click OK and no other message boxes popup

### Expected Behavior

A message box that says Child1View has been disposed and then another message box that says Child2View has been disposed.

This is because RegionManager.UpdateRegions() at all in the close process. I feel it should fire after a call to Region.Remove(view), either all of the time or if it knows that view has regions within in it, it will then always be up to date. I am currently making a manual call to that in my close method but feel that this should be the default behavior as this has taken some investigation.

### Actual Behavior

No other message boxes popup, you have to click the button to Navigate to View with Child Views again, for this to force the clear up of the views.

### Basic Information

- Version with issue: 8.1.97
- Last known good version: Unknown
- Xamarin.Forms version: N/A
- IDE: Visual Studio 2019

### Screenshots

N/A

### Reproduction Link

[PrismChildRegions.zip](https://github.com/PrismLibrary/Prism/files/7454268/PrismChildRegions.zip)

**Comment by brianlagunas** [comment]
This is not a bug. `IDestructible` is only called when a view is removed from a region. If you have a view with multiple regions, or even regions within regions, and you close the parent view, the child regions still contain the views. Therefore nothing has been removed from the region. Meaning the `IDestuctible` interface will never be called.

Prism will never be able to know how many region exist in a view, what views exist in those regions, if and when regions become part of the visual tree, when that parent view is removed from it's parent container (assuming it even has a parent container), the lifecycle of those objects (if thy are reused or not), and all the other unknown variables when building applications.

You are responsible for managing your regions and the views within those regions. That is why Prism provides its Region and Navigation APIs.

Based on your current approach, the best way to clean up your child regions would be to handle the `IDestructible` interface on the parent view code behind and manually remove any views within those regions. That would then call the interface on all the nested views within those child regions.

Also, keep in mind that `IDestructible` is not `IDisposable` and should not be used in the same way.

**Comment by james1301** [comment]
@brianlagunas I am not saying the bug has anything to do with calling IDestructible, it should however be called in my sample, in what I thought was a nice way of clearing up child views and highlights the actual problem I am trying to show. That is when removing a view, regions are not updated, when it seems a likely occurrence in more complex applications and/or applications that are trying to be more composite.

**Comment by james1301** [comment]
This problem also means that the ClearChildViewsRegionBehavior does not work very well.

**Comment by brianlagunas** [comment]
`ClearChildViewsRegionBehavior` works as designed.  It is a behavior that removes the RegionManager attached property of all the views in a region once the RegionManager property of a **region** becomes null. Not just when any view in a region is removed.

When a view containing a region is removed from the visual tree, views within that region are not removed. This acts just like every other control in WPF. If you close a WPF Window, all of the elements within that window are not removed from the visual tree.

If you want to actually **remove** any views from a nested region, you can do so manually by calling the `Region.RemoveAll` method when the parent view is removed from it's region.

If you want to force all the RegionBehaviors to update their region managers immediately so that you can use your custom region behavior immediately then just call `RegionManager.UpdateRegions()`. Otherwise, this process will happen when Prism gets to it.

**Comment by james1301** [comment]
I get what you're saying but I'm not sure that works as well as it could do in practice. Until those regions are removed it keeps all of the views in memory. So either it relies on the user navigating somewhere else or the developer manually calls that update.

I've decided to manually call it as to cause less confusion for other developers that those views still exist in memory, because I think it requires a bit of knowledge to know that they are still there and that is OK because they will get sorted in the next navigation.

The ClearChildViewsRegionBehavior un-links the views but it is likely you will want to sort out the ViewModel too which is why I've just replaced it with a call to deactivate the view rather than just to null the region manager. That way all of those other good things like IDestructible etc will get fired etc. As in practice there are probably going to be things to dispose of, events to unsubscribe etc.

For the effort of one call to RegionManger.UpdateRegions and possibly to change the ClearChildViews behavior to deactivate instead adds a lot out of the box for Prism. It would have saved me a fair bit of time, of what is this behavior doing (confused with it not doing something straight away)? Have I set it in the right place? Does it go on the parent view? Parent Region? No each child view. Years ago I did something else instead, because I was unsure what it was doing or how it was supposed to be used. And revisiting it on this project I thought I'd try and get to the bottom of it, because Prism is with most aspects really good with covering a lot of the MVVM bases.
